File: core/src/nl/t64/game/rpg/components/quest/QuestTask.java
package nl.t64.game.rpg.components.quest;

import com.badlogic.gdx.utils.GdxRuntimeException;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Getter;
import nl.t64.game.rpg.Utils;

import java.util.Map;


@Getter
public class QuestTask {

    QuestType type;
    private String taskPhrase;
    private Map<String, Integer> target;
    @JsonProperty("isOptional")
    private boolean isOptional;
    private boolean isComplete;
    private boolean isFailed;
    private boolean isQuestFinished;

    public QuestTask(String taskPhrase) {
        this();
        this.type = QuestType.NONE;
        this.taskPhrase = taskPhrase;
    }

    private QuestTask() {
        this.target = Map.of();
        this.isOptional = false;
        this.isComplete = false;
        this.isFailed = false;
        this.isQuestFinished = false;
    }

    public String toString() {
        if (type.equals(QuestType.NONE)) {
            return System.lineSeparator() + System.lineSeparator() + System.lineSeparator() + taskPhrase;
        } else if (isFailed) {
            return "x  " + taskPhrase;
        } else if (isQuestFinished) {
            return "v  " + taskPhrase;
        } else if (type.equals(QuestType.RETURN)) {
            return "     " + taskPhrase;
        } else if (isCompleteForReturn()) {
            return "v  " + taskPhrase;
        } else {
            return "     " + taskPhrase;
        }
    }

    void forceFinished() {
        isQuestFinished = true;
        if (isOptional && !isComplete) {
            isFailed = true;
        }
    }

    void setComplete() {
        switch (type) {
            case ITEM_DELIVERY -> {
                target.forEach((itemId, amount) -> Utils.getGameData().getInventory().autoRemoveItem(itemId, amount));
                isComplete = true;
            }
            case DISCOVER, MESSAGE_DELIVERY -> isComplete = true;
            case CHECK -> setCheckTypePossibleComplete();
            default -> throw new IllegalCallerException("Only possible to complete a DISCOVER, CHECK or DELIVERY task.");
        }
    }

    private void setCheckTypePossibleComplete() {
        if (target.isEmpty()) {
            isComplete = true;
            return;
        }

        if (doesInventoryContainsTarget()) {
            isComplete = true;
        }
    }

    void removeTargetFromInventory() {
        Utils.getGameData().getInventory().autoRemoveItem(getTargetEntry().getKey(),
                                                          getTargetEntry().getValue());
    }

    boolean hasTargetInInventory() {
        return Utils.getGameData().getInventory().contains(target);
    }

    boolean isCompleteForReturn() {
        return switch (type) {
            case RETURN -> true;
            case FETCH -> doesInventoryContainsTarget();
            case DISCOVER, CHECK, MESSAGE_DELIVERY, ITEM_DELIVERY -> isComplete;
            default -> throw new GdxRuntimeException(String.format("No %s task for now.", type));
        };
    }

    private boolean doesInventoryContainsTarget() {
        return Utils.getGameData().getInventory().hasEnoughOfItem(getTargetEntry().getKey(),
                                                                  getTargetEntry().getValue());
    }

    private Map.Entry<String, Integer> getTargetEntry() {
        return target.entrySet().iterator().next();
    }

}

File: core/src/nl/t64/game/rpg/components/quest/QuestTask.kt
package nl.t64.game.rpg.components.quest

import nl.t64.game.rpg.Utils.gameData


class QuestTask(
    val taskPhrase: String = "",
    val type: QuestType = QuestType.NONE,
    val target: Map<String, Int> = emptyMap(),
    val isOptional: Boolean = false,
) {
    var isComplete: Boolean = false
    var isFailed: Boolean = false
    var isQuestFinished: Boolean = false

    override fun toString(): String {
        return when {
            type == QuestType.NONE -> System.lineSeparator() + System.lineSeparator() + System.lineSeparator() + taskPhrase
            isFailed -> "x  $taskPhrase"
            isQuestFinished -> "v  $taskPhrase"
            type == QuestType.RETURN -> "     $taskPhrase"
            isCompleteForReturn() -> "v  $taskPhrase"
            else -> "     $taskPhrase"
        }
    }

    fun forceFinished() {
        isQuestFinished = true
        if (isOptional && !isComplete) {
            isFailed = true
        }
    }

    fun setComplete() {
        when (type) {
            QuestType.ITEM_DELIVERY -> {
                target.forEach { (itemId, amount) -> gameData.inventory.autoRemoveItem(itemId, amount) }
                isComplete = true
            }
            QuestType.DISCOVER,
            QuestType.MESSAGE_DELIVERY -> isComplete = true
            QuestType.CHECK -> setCheckTypePossibleComplete()
            else -> throw IllegalArgumentException("Only possible to complete a DISCOVER, CHECK or DELIVERY task.")
        }
    }

    private fun setCheckTypePossibleComplete() {
        if (target.isEmpty()) {
            isComplete = true
            return
        }

        if (doesInventoryContainsTarget()) {
            isComplete = true
        }
    }

    fun removeTargetFromInventory() {
        gameData.inventory.autoRemoveItem(getTargetEntry().key, getTargetEntry().value)
    }

    fun hasTargetInInventory(): Boolean {
        return gameData.inventory.contains(target)
    }

    fun isCompleteForReturn(): Boolean {
        return when (type) {
            QuestType.RETURN -> true
            QuestType.FETCH -> doesInventoryContainsTarget()
            QuestType.DISCOVER,
            QuestType.CHECK,
            QuestType.MESSAGE_DELIVERY,
            QuestType.ITEM_DELIVERY -> isComplete
            else -> throw IllegalArgumentException("No '$type' task for now.")
        }
    }

    private fun doesInventoryContainsTarget(): Boolean {
        return gameData.inventory.hasEnoughOfItem(getTargetEntry().key, getTargetEntry().value)
    }

    private fun getTargetEntry(): Map.Entry<String, Int> {
        return target.entries.iterator().next()
    }

}
